File: backend-python/app/services/distribution_service.py
from typing import List, Dict, Any
import math

def distribute_items(categories: List[Dict[str, Any]], num_boxes: int, stores: List[Dict[str, Any]]):
    available_categories = [c for c in categories if c.get('total_entries', 0) > 0]
    total_items = sum(c.get('total_entries', 0) for c in available_categories)

    if total_items == 0:
        raise ValueError("Nenhum produto cadastrado para distribuir.")

    if num_boxes <= 0:
        raise ValueError("Número de caixas deve ser maior que zero.")

    if num_boxes > total_items:
        raise ValueError(f"Impossível criar {num_boxes} caixas com apenas {total_items} itens.")

    # Inicializar caixas vazias
    boxes = [{"box_number": i + 1, "items": {}} for i in range(num_boxes)]

    # Passo 1: Distribuição Round-Robin com Rotação
    for cat_index, category in enumerate(available_categories):
        total_entries = category['total_entries']
        base = total_entries // num_boxes
        remainder = total_entries % num_boxes
        
        # Offset rotativo por categoria
        offset = cat_index % num_boxes

        for i in range(num_boxes):
            rotated_index = (i + offset) % num_boxes
            bonus = 1 if rotated_index < remainder else 0
            quantity = base + bonus
            
            if quantity > 0:
                boxes[i]["items"][category["id"]] = quantity

    # Passo 2: Atribuição de Lojas
    for i, box in enumerate(boxes):
        store = stores[i % len(stores)]
        box["assigned_store_id"] = store["id"]
        box["assigned_store_name"] = store["name"]

    # Passo 3: Warnings
    warnings = []
    for cat in available_categories:
        if cat['total_entries'] < num_boxes:
            missing = num_boxes - cat['total_entries']
            warnings.append(f'"{cat["name"]}" tem apenas {cat["total_entries"]} itens — {missing} caixa(s) ficarão sem esta categoria.')

    return {"boxes": boxes, "warnings": warnings}

def suggest_box_count(categories: List[Dict[str, Any]], stores_count: int):
    ITEMS_PER_BOX_IDEAL = 15
    total_items = sum(c.get('total_entries', 0) for c in categories)
    
    if total_items == 0:
        return {"suggested": stores_count, "reasoning": "Sem itens no inventário."}
        
    by_capacity = math.ceil(total_items / ITEMS_PER_BOX_IDEAL)
    suggestion = max(stores_count, by_capacity)

    return {
        "suggested": suggestion,
        "reasoning": f"{stores_count} lojas, {total_items} itens → mínimo {stores_count} caixas (1/loja), ideal {by_capacity} caixas (~{ITEMS_PER_BOX_IDEAL} itens/caixa)"
    }

File: backend-node/src/services/distributionService.js
const { v4: uuidv4 } = require('uuid');

/**
 * Algoritmo de Distribuição Justa (Round-Robin com Rotação)
 */
function distributeItems(categories, numBoxes, stores) {
  const availableCategories = categories.filter(c => c.total_entries > 0);
  const totalItems = availableCategories.reduce((sum, c) => sum + (c.total_entries || 0), 0);

  if (totalItems === 0) {
    throw new Error('Nenhum produto cadastrado para distribuir.');
  }

  if (numBoxes <= 0) {
    throw new Error('Número de caixas deve ser maior que zero.');
  }

  if (numBoxes > totalItems) {
    throw new Error(`Impossível criar ${numBoxes} caixas com apenas ${totalItems} itens.`);
  }

  // Inicializar caixas vazias
  const boxes = Array.from({ length: numBoxes }, (_, i) => ({
    box_number: i + 1,
    items: {} // category_id -> target_quantity
  }));

  // Passo 1: Distribuir itens por categoria
  availableCategories.forEach((category, catIndex) => {
    const base = Math.floor(category.total_entries / numBoxes);
    const remainder = category.total_entries % numBoxes;

    // Offset rotativo para evitar que a Caixa 1 sempre ganhe o resto
    const offset = catIndex % numBoxes;

    for (let i = 0; i < numBoxes; i++) {
      const rotatedIndex = (i + offset) % numBoxes;
      const bonus = rotatedIndex < remainder ? 1 : 0;
      const quantity = base + bonus;

      if (quantity > 0) {
        boxes[i].items[category.id] = quantity;
      }
    }
  });

  // Passo 2: Atribuição de Lojas
  // (Nota: assumimos que stores já vem ordenado por quem tem menos caixas se for recálculo,
  // ou apenas a lista total se for criação inicial)
  boxes.forEach((box, i) => {
    const store = stores[i % stores.length];
    box.assigned_store_id = store.id;
    box.assigned_store_name = store.name;
  });

  // Passo 3: Gerar Warnings
  const warnings = [];
  availableCategories.forEach(cat => {
    if (cat.total_entries < numBoxes) {
      const missing = numBoxes - cat.total_entries;
      warnings.push(`"${cat.name}" tem apenas ${cat.total_entries} itens — ${missing} caixa(s) ficarão sem esta categoria.`);
    }
  });

  return { boxes, warnings };
}

function suggestBoxCount(categories, storesCount) {
  const ITEMS_PER_BOX_IDEAL = 15;
  const totalItems = categories.reduce((sum, c) => sum + (c.total_entries || 0), 0);
  
  const byCapacity = Math.ceil(totalItems / ITEMS_PER_BOX_IDEAL);
  const suggestion = Math.max(storesCount, byCapacity);

  return {
    suggested: suggestion,
    reasoning: `${storesCount} lojas, ${totalItems} itens → mínimo ${storesCount} caixas (1/loja), ideal ${byCapacity} caixas (~${ITEMS_PER_BOX_IDEAL} itens/caixa)`
  };
}

module.exports = {
  distributeItems,
  suggestBoxCount
};
